### ISSUE TITLE
Inference silently returns all-NaN logits — 10% of code documents on one checkpoint, no guard anywhere

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#229]
Found by accident while running a depth sweep across the checkpoint ladder (#226): the
`best/3121279` checkpoint (opt step 24,385, 3.20B tokens) returns **all-NaN logits on ~10% of
code documents** under the f16 compute policy. Nothing warns; the number just becomes NaN.

## Measured

```
checkpoint                corpus         NaN rows / 300
best/3121279 (3.20B)      codeparrot     29/300   (9.67%)
best/3121279 (3.20B)      fineweb-edu     0/300   (0.00%)
champion 3899391 (3.99B)  codeparrot      0/300   (0.00%)
champion 3899391 (3.99B)  finemath        0/300   (0.00%)
```

Depth 1 is enough to trigger it, so **the refine loop is not involved** — it is the encoder or
the head. Every parameter array in the checkpoint is finite (150/150 checked), so the stored
weights are not corrupt.

It is purely a compute-precision effect:

```
                        f16                          f32
3.20B, offending row    window1 all-NaN              window1 finite, max|logit| 23.1
champion, same row      finite, max|logit| 30.0      finite, max|logit| 30.1
```

`FORCE_F32_COMPUTE=1` makes it go away. Note the champion runs *hotter* (max|logit| 30 vs 23)
and does not overflow, so peak logit magnitude is not the predictor — something in an
intermediate activation is.

## Why this matters more than one bad checkpoint

**The inference path has no non-finite guard.** The training path does — `#199`'s scaler
detected and skipped 14,163 non-finite micro-steps during the base run, which is exactly this
class of event being caught and handled. Inference has no equivalent: a NaN just flows into
whatever consumes it.

Consequences already visible or plausible:

- The depth sweep at this rung returned `nan` for every statistic. That is the **good** case —
  it failed loudly. The dangerous case is any consumer that averages, masks, or `nan_to_num`s
  and reports a plausible number instead.
- This checkpoint was used for the yardstick ladder (LAMBADA 9.45%, held-out ppl 38.76). Both
  are prose-based and measured 0% NaN here, so those numbers are probably unaffected — but
  that is **luck, not design**. Nothing in that path would have told us either way.
- Any generation from an affected checkpoint on code input would produce garbage with no error.

## The wider question this raises

`COMPUTE_DTYPE = float16` is a permanent policy (Turing card, no bf16) and it is not
negotiable. What is negotiable is that we currently have **no idea how common this is**. One
intermediate checkpoint fails on a tenth of one corpus; the champion does not. We have never
checked any other stored checkpoint, and we would not have found this one if a sweep had not
happened to read code documents through it.

## Proposed

1. **Guard the inference path.** A non-finite check on logits in `trm/infer.py` and in the
   eval helpers, raising or reporting rather than propagating. Cheap, and it converts a silent
   corruption into a visible failure.
2. **A NaN-incidence smoke** in `instruments/` — N documents per corpus through a checkpoint at
   f16, report the rate. Run it against any checkpoint before that checkpoint is used for a
   measurement or archived as notable.
3. **Find the overflow.** Instrument activations layer-by-layer on a known-offending row at f16
   to locate where it blows up. That is the only route to knowing whether this is a
   this-checkpoint accident or a latent property of the architecture at f16.
4. Add the rate to the model-card template, so "0% NaN on N documents per corpus" is recorded
   for anything we keep.

Item 1 is the one worth doing immediately; the rest can queue.

## Repro

```bash
MODEL_ARCH=refiner PYTHONPATH=. venv/bin/python -c "
import jax.numpy as jnp
from trm.config import MAX_SEQ_LEN
from trm.data.loaders import TextDataGenerator
from trm.runtime.restore import build_model, _restore_into
m,_ = _restore_into(build_model(), 'runs/run_20260813_214725/checkpoints/best')
g = TextDataGenerator('runs/data/pretrain/codeparrot'); g.skip_count = 3_000_000
bad = sum(not bool(jnp.all(jnp.isfinite(
          m(jnp.asarray(g.get_batch(1)[0][:, :MAX_SEQ_LEN]), depth=1,
            training=False, new_document=True).logits))) for _ in range(300))
print(f'{bad}/300 NaN')"
```

(The `best/` manager resolves to step 3121279, the affected checkpoint.)

## Lane

`trm/infer.py` + `instruments/`. Fully CPU-reproducible for the guard and the smoke; item 3
wants the GPU only for speed.

**Comment by MatthewLacerda2** [comment]
Item 1 (the guard) is in **PR #234**. Items 2–4 stay open here.

Measured against this checkpoint rather than synthetic arrays — 300 codeparrot documents,
depth 1, guard verdict compared to ground truth per row:

```
                 non-finite somewhere   at last position   guard fired   MISSED
best/3121279          13/300                 13/300            13          0
champion 3899391       0/300                  0/300             0          0
```

Two results worth recording here, one of which contradicts what I expected:

## 1. The NaN is global, not positional

"Somewhere" and "at the last position" are the **same 13 documents**. An affected document is
non-finite at *every* one of its 512 positions, not at the tail or at a scattered few.

That matters for item 3 (locating where it blows up), because it rules out a per-position
overflow and points at something the whole window shares. It is also the same signature as
**#233**, where one out-of-vocab token id turns all 512 positions non-finite — including
positions causally *before* the bad token, so it cannot be travelling through attention. Two
different triggers producing the same whole-window spread raises the odds they share a
mechanism: a norm, or the shared refine state.

`instruments/latents.py` (#225) answers this directly — capture the trajectory on one
affected document and read off whether `z_0` (the encoder output) is already non-finite or
whether a specific refine pass introduces it. That is a better first customer for the new
instrument than looking for structure and hoping.

## 2. The rate did not reproduce, and I cannot yet say why

**13/300 here, 29/300 as filed.** Same checkpoint, same corpus, same `skip_count = 3_000_000`,
same depth. `TextDataGenerator` seeds from `np.random.default_rng(DATA_SEED)`, so the document
set should be identical, and two separate runs here both returned 13.

The most likely remaining difference is the **backend**: this issue notes the repro is
CPU-reproducible, and these runs were on the GPU. f16 overflow incidence is precisely the kind
of thing that differs between the CPU emitter and the CUDA one — which, if true, is itself
worth knowing, because it would mean the failure rate depends on where you serve.

Not resolved. Flagged rather than smoothed over, since the 9.67% figure is currently the
headline number on this issue and it may be backend-specific. The cheap test is to re-run the
same 300 documents under `FORCE_F32_COMPUTE=1` and on the CPU emitter and compare all three;
that folds naturally into item 3.

None of this affects the guard: its verdict is computed per row against ground truth on
whatever rows appear, and it missed none.

**Comment by MatthewLacerda2** [comment]
## Root cause located — and this issue already said where to look

> "Depth 1 is enough to trigger it, so **the refine loop is not involved** — it is the encoder
> or the head. Every parameter array in the checkpoint is finite (150/150 checked), so the
> stored weights are not corrupt."

That was right. Measured with the new trajectory instrument (#225), on 4 held-out documents per
corpus, both surviving checkpoints:

```
                       max|encoder output|          max|state| through 8 passes
fineweb-edu                      47.7                        48
codeparrot                   65,120                      56,032
                                                       (f16 max = 65,504)
```

**The encoder's output on code reaches 65,120 — 99.4% of the f16 ceiling, and 1,365× the prose
value.** Identical on `best/3121279` (3.20B: 65,440) and the champion (3.99B: 65,120), so it is
a stable property of the architecture on this corpus, not drift.

That accounts for every measured feature of this bug:

| observation | explained by |
|---|---|
| code only, prose never | prose runs at 48; code runs at the ceiling |
| depth 1 triggers it | z_enc is computed before any refinement |
| non-finite at **every** position | z_enc feeds the whole window |
| ~10% of documents, not all | only documents that push past the last 0.6% overflow |
| weights all finite | the weights are fine; the *activations* are not |

Confirmed downstream too: the gate pre-activation on codeparrot pass 1 reaches **74,292.9** when
the same matmul is accumulated in f32, against a 65,504 ceiling, producing 243 non-finite values
in the f16 path. So the overflow is genuine rather than an accumulation artifact.

## Why the champion looks clean and is not

The champion returned 0/300 NaN on codeparrot in the acceptance run above, so it serves fine
today. But it is running at 65,120 of 65,504 — **0.6% of headroom**. That is not a fixed bug; it
is the same bug that has not yet been handed an input slightly worse than the 300 we sampled.
Any change that nudges activations up — different data, a longer context, a resumed run — walks
straight into it.

## What follows

1. **Item 1 (the guard) is now doubly justified** and is in PR #234: the failure is one rounding
   away on the model we actually serve.
2. **Item 3 (locate where it blows up) is done** — it is the encoder output, and it did not need
   a bisect.
3. **New, and more important than the rest**: `max|z_enc|` should be telemetry. The activation
   scale that decides whether this model can be served in f16 at all has never been measured
   during a run. That is `tools` and cheap — one scalar per validation batch.
4. The remaining research question is *why* the encoder produces 10³–10⁴ magnitude activations
   at all. The residual stream is unnormalized between blocks (RMSNorm sits at `time_norm` and
   `out_norm`, not on the residual path), so there is nothing bounding it. Whether that wants
   fixing is an `ideas` question and needs a matched ablation, not a patch.

Retraction from my earlier comment on #225: I attributed this to the refine gate saturating.
The gate has `max|w| = 0.10` and is simply being fed an oversized state; it is a symptom.
Corrected on #235, which is now retitled to the encoder.

**Comment by MatthewLacerda2** [comment]
Closed by #234. Items 1 and 3 are done — the guard landed, and the overflow was located at the encoder output (65,120 of f16's 65,504 on code, both surviving checkpoints), which also answered item 3's real question: it is a **latent property of the architecture at f16**, not a this-checkpoint accident.

Items 2 and 4 are unfinished and now belong with the cause rather than the symptom, so they move to **#235**: a NaN-incidence smoke and a model-card field are both really "record the activation headroom of anything we keep". Tracking them there so this does not sit open on two small chores whose rationale has changed.
